provers/evm-prover/src/main.rs

```diff
@@ -12,55 +21,202 @@ pub mod prover {
         };
 
         Ok(Response::new(response))
     }
 
     async fn prove_state_transition(
         &self,
-        _request: Request<ProveStateTransitionRequest>,
+        request: Request<ProveStateTransitionRequest>,
     ) -> Result<Response<ProveStateTransitionResponse>, Status> {
-        Err(Status::unimplemented(
-            "State transition proofs not yet implemented",
-        ))
+        println!("Got state transition request: {:?}", request);
+
+        let inner_request = request.into_inner();
+
+        // Get the latest height from EVM rollup.
+        let latest_height = self.get_latest_height().await?;
+
+        // Get the trusted height from groth16 client.
+        let trusted_height = self
+            .get_trusted_height(inner_request.client_id.as_str())
+            .await?;
+
+        println!(
+            "proving from height {:?} to height {:?}",
+            latest_height, trusted_height
+        );
+
+        if latest_height.as_u64() <= trusted_height {
+            return Err(Status::unimplemented(
+                "Trusted height is greater than latest height",
+            ));
+        }
+
+        let mut inputs = vec![];
+        for height in trusted_height + 1..=latest_height.as_u64() {
+            let (inclusion_height, blob_commitment) =
+                get_inclusion_height(self.indexer_url.clone(), height)
+                    .await
+                    .unwrap();
+            let client_executor_input = generate_client_input(
+                self.evm_rpc_url.clone(),
+                height,
+                &self.genesis,
+                self.custom_beneficiary.as_ref(),
+                self.opcode_tracking,
+            )
+            .await
+            .unwrap();
+            let input = BlockProverInput {
+                inclusion_height,
+                client_executor_input,
+                rollup_block: vec![],
+            };
+            inputs.push(input);
+        }
+
+        // Generate proofs and collect verifying keys
+        let mut aggregation_inputs = vec![];
+        for input in inputs {
+            let (proof, vk) = self.prover.generate_proof(input).await.unwrap();
+            aggregation_inputs.push(AggregationInput {
+                proof,
+                vk: vk.clone(),
+            });
+        }
+
+        let aggregation_output = self
+            .prover
+            .aggregate_proofs(aggregation_inputs)
+            .await
+            .unwrap();
+
+        let response = ProveStateTransitionResponse {
+            proof: bincode::serialize(&aggregation_output.proof.proof).unwrap(),
+            public_values: aggregation_output.proof.public_values.to_vec(),
+        };
+
+        Ok(Response::new(response))
     }
 
     async fn prove_state_membership(
         &self,
         _request: Request<ProveStateMembershipRequest>,
     ) -> Result<Response<ProveStateMembershipResponse>, Status> {
+        // TODO: Implement membership proofs later
```

Comment:
could be exported into an issue if we care about this post demo

Reply:
Seems like #222 covers that.